### PR TITLE
[macOS] New macOS Trackers animation

### PR DESCRIPTION
**Pull request by afterxleep**
Task/Issue URL: https://app.asana.com/1/137249556945/project/1204099484721401/task/1210870990447536?focus=true

### Description

Adds new tracker blocking animation for the Privacy Shield button in the macOS navigation bar. 

### Demo
CleanShot 2025-11-28 at 15 19 00

**Key Changes**:
- Implemented new tracker blocking animation with improved timing

### Testing Steps

1. Open DuckDuckGo browser on macOS
2. Navigate to a website with trackers
3. Observe the Privacy Shield animation when trackers are blocked
4. Test with privacy protections disabled to verify icon displays correctly

### Impact and Risks
- **Edge cases**: Tested with 0, 1, and multiple trackers blocked scenarios

---

###### Internal references:

Definition of Done | Engineering Expectations | Tech Design Template

---

> [!NOTE]
> Implements a prioritized badge notification system (trackers → shield → cookies) in the macOS address bar with new Lottie shield animations, tab-aware handling, and comprehensive tests.
> 
> - **macOS UI/Logic**:
>   - **Badge Notifications**: Introduces `NavigationBarBadgeAnimator` with a priority queue (trackers high, cookies low), auto-processing control, and tab-switch-aware cancellation.
>   - **Controller Integration**: Updates `AddressBarButtonsViewController` to enqueue notifications, play shield animation after tracker notices, and manage hover/visibility states; refines mouse-over behavior and layout.
>   - **Views**: Adds `NavigationBarBadgeAnimationView`, `BadgeNotificationContainerView`, `BadgeNotificationContentView`, and related animation models; replaces cookie-specific views with generic badge components.
> - **Animations/Assets**:
>   - Adds/updates Lottie resources for shield/trackers (including `shield.new.json`) and aligns asset selection via `PrivacyShieldAddressBarStyleProviding`.
>   - Tweaks storyboard constraints for badge area alignment.
> - **Localization**:
>   - Adds strings and pluralization for "Trackers Blocked" and cookie notifications; minor localization fixes.
> - **Tests**:
>   - Adds unit tests for animator, animation view, and controller animation state; updates integration tests for shield animation visibility.
> - **Misc**:
>   - Minor dependency/version and `.gitignore` updates.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4f715a99eb4cb196abd82f07554a4ba8cead2d16. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift [hunk 1]
```diff
@@ -1952,11 +1990,16 @@ final class AddressBarButtonsViewController: NSViewController {
...
+                // Hide the Lottie shield animations when hover animation is visible
+                // to prevent overlap and misalignment
                 if isAnimationViewVisible {
-                    self?.stopAnimations(trackerAnimations: false, shieldAnimations: true, badgeAnimations: false)
+                    self.shieldAnimationView?.isHidden = true
+                    self.shieldDotAnimationView?.isHidden = true
                 } else {
-                    self?.updatePrivacyEntryPointIcon()
+                    // Restore shield visibility when hover animation ends
+                    self.updatePrivacyEntryPointIcon()
```
[the remark this answers is not in the document]
Comment (afterxleep) [reply macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift:2328]: thats fine, animation covers the shield icon when playing

File: macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift [hunk 2]
```diff
@@ -431,27 +435,31 @@ final class AddressBarButtonsViewController: NSViewController {
...
-            DispatchQueue.main.asyncAfter(deadline: .now() + 0.6) {
-                if self.tabViewModel?.tab == queuedNotification.selectedTab {
-                    self.showBadgeNotification(queuedNotification.animationType)
-                } else {
-                    self.buttonsBadgeAnimator.queuedAnimation = nil
-                }
-            }
-        }
+    /// Shows a tracker notification with the count of trackers blocked
+    /// - Parameter count: Number of trackers blocked
```
Comment (jaceklyp) [inline macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift:496]: this piece of doc seems a bit of redundant

File: macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift [hunk 3]
```diff
@@ -220,6 +221,9 @@ final class AddressBarButtonsViewController: NSViewController {
     private var trackerAnimationTriggerCancellable: AnyCancellable?
     private var privacyEntryPointIconUpdateCancellable: AnyCancellable?
 
+    private var lastNotificationType: NavigationBarBadgeAnimationView.AnimationType?
+    private var trackerNotificationShownForURL: URL?
```
Comment (jaceklyp) [inline macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift]: maybe if we already use lastNotificationType this one should be called lastNotifiedURL: URL?
alternatively you could use "previous" prefix

File: macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift [hunk 4]
```diff
@@ -2057,7 +2113,61 @@ extension AddressBarButtonsViewController {
...
+                self.hasShieldAnimationCompleted = true
+
+                // After shield animation completes, process next queued notification (like cookies)
+                self.buttonsBadgeAnimator.processNextAnimation()
+                self.playPrivacyInfoHighlightAnimationIfNecessary()
+            }
+
+            // Return early - don't process next animation yet, wait for shield to complete
+            return
+        }
```
Comment (jaceklyp) [inline macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift:2498]: I have a hard time understanding this piece of code - it's long, quite heavy in logic - e.g. there are many returns and some without queue not advancing, I'm not sure if this is valid or not?

Could you also extract code that runs the animation into a separate `func playShieldAnimation(for url: URL)` not to combine animation code with our business logic?

Comment (afterxleep) [reply macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift:2498]: Yea, the early returns are intentional (no URL, animator busy, not Https site), which should not process the animation.

Good call on the method.  Extracted.

File: macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift [hunk 5]
```diff
@@ -2243,8 +2301,48 @@ extension AddressBarButtonsViewController {
...
+
+        playShieldAnimation(for: url)
+    }
+
+    private func playShieldAnimation(for url: URL) {
+        let endFrame = shieldAnimationView.animation?.endFrame ?? 0
+        shieldAnimationView.play(fromFrame: 1, toFrame: endFrame, loopMode: .playOnce) { [weak self] finished in
+            guard finished, let self = self else { return }
```
[the remark this answers is not in the document]
Comment (afterxleep) [reply macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift:2480]: This is fine.  Lottie file is local and visibility is covered by E2E tests.

File: macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift [hunk 6]
```diff
@@ -214,6 +215,15 @@ final class AddressBarButtonsViewController: NSViewController {
...
             updateButtons()
             stopHighlightingPrivacyShield()
+            if isTextFieldEditorFirstResponder {
+                // Hide shield when address bar is focused
+                updatePrivacyEntryPointIcon()
+            } else {
+                // Restore shield when address bar loses focus
+                hasShieldAnimationCompleted = false
+                shieldAnimationView?.currentFrame = 1
+                updatePrivacyEntryPointIcon()
```
[the remark this answers is not in the document]
Comment (afterxleep) [reply macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift:222]: this is expected behavior

File: macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift [hunk 7]
```diff
@@ -516,8 +532,15 @@ final class AddressBarButtonsViewController: NSViewController {
     private func subscribeToSelectedTabViewModel() {
         tabCollectionViewModel.$selectedTabViewModel.sink { [weak self] tabViewModel in
             guard let self else { return }
-
-            stopAnimations()
+            // Stop visual animations but let the animator handle queue management
+            // The animator's handleTabSwitch preserves animations for the current tab
+            stopAnimations(badgeAnimations: false)
```
[the remark this answers is not in the document]
Comment (afterxleep) [reply macOS/DuckDuckGo/NavigationBar/View/AddressBarButtonsViewController.swift:537]: this is expected behavior
